src/SubHask/Algebra/Vector.hs

```diff
@@ -1497,6 +1515,130 @@ instance
 
 --------------------------------------------------------------------------------
 
+-- | Accelerate based Vector
+-- | A.Acc is an accelreate computation, A.Array A.DIM1 a is a one dimensional array
+
+newtype ACCVector (bknd::Backend) (n::k) a = ACCVector (A.Acc (A.Array A.DIM1 a))
+
+type instance Scalar (ACCVector bknd n r) = Scalar r
+type instance Logic (ACCVector bknd n r) = Logic r
+
+type instance ACCVector bknd m a >< b = Tensor_ACCVector (ACCVector bknd m a) b
+type family Tensor_ACCVector a b where
+    Tensor_ACCVector (ACCVector bknd n r1) (ACCVector bknd m r2) = ACCVector bknd n r1 +> ACCVector bknd m r2
+    Tensor_ACCVector (ACCVector bknd n r1) r1 = ACCVector bknd n r1 -- (r1><r2)
+
+type ValidACCVector bknd n a = ( (ACCVector (bknd::Backend) n a><Scalar a)~ACCVector (bknd::Backend) n a, Prim a, A.Elt a, A.IsNum a, IsScalar a)
+
+type instance Index (ACCVector bknd n r) = Int
+type instance Elem (ACCVector bknd n r) = Scalar r
+type instance SetElem (ACCVector (bknd::Backend) n r) b = ACCVector (bknd::Backend) n b
+
+
+type instance Actor (ACCVector (bknd::Backend) n r) = Actor r
+
+
+mkAccVectorFromList :: A.Elt a => [a] -> ACCVector bknd (n::Symbol) a
+mkAccVectorFromList l = let
+    len = P.length l
+  in ACCVector (A.use (A.fromList (A.Z A.:.len) l))
+
+mkAccVector :: (A.Elt a, ValidSVector (n::Symbol) a) => SVector (n::Symbol) a -> ACCVector (bknd::Backend) (n::Symbol) a
+mkAccVector v @(SVector_Dynamic fp off n) = let
+  arr = A.fromList (A.Z A.:. n) $ unsafeInlineIO $ go (n-1) []
+  go (-1) xs = return $ xs
+  go i    xs = withForeignPtr fp $ \p -> do
+      x <- peekElemOff p (off+i)
+      go (i-1) (x:xs)
+  in ACCVector (A.use arr)
+
+
+-- acc2SVector fails with:
```

Comment:
You probably need to add `Scalar a ~ Scalar (Scalar a)` to the list of constraints for `ValidAccVector`